### PR TITLE
Changing new custom serialization API based on API review feedback

### PR DESCRIPTION
**Pull request by FabianMeiswinkel**
# Description
Changing new custom serialization API based on API review feedback

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/CHANGELOG.md [hunk 1]
```diff
@@ -3,7 +3,7 @@
 ### 4.59.0-beta.1 (Unreleased)
 
 #### Features Added
-* Added public APIs `getCustomeSerializer` and `setCustomSerializer` to allow customers to specify custom payload transformations or serialization settings. - See [PR 38997](https://github.com/Azure/azure-sdk-for-java/pull/38997)
+* Added public APIs `getCustomeSerializer` and `setCustomSerializer` to allow customers to specify custom payload transformations or serialization settings. - See [PR 38997](https://github.com/Azure/azure-sdk-for-java/pull/38997) and [PR 39933](https://github.com/Azure/azure-sdk-for-java/pull/39933)
```
Comment (tvaron3) [inline sdk/cosmos/azure-cosmos/CHANGELOG.md:6]:
```suggestion
* Added public APIs `getCustomItemSerializer` and `setCustomItemSerializer` to allow customers to specify custom payload transformations or serialization settings. - See [PR 38997](https://github.com/Azure/azure-sdk-for-java/pull/38997) and [PR 39933](https://github.com/Azure/azure-sdk-for-java/pull/39933) 
```

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/NonIdempotentWriteRetryOptions.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.azure.cosmos;
+
+/**
+ * Class to define the options for non-idempotent write operations
+ */
+public final class NonIdempotentWriteRetryOptions {
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/NonIdempotentWriteRetryOptions.java:8]: Most of our `options` classes are in `com.azure.cosmos.models` package. Any specific reason we are putting this outside?

File: sdk/cosmos/azure-cosmos/CHANGELOG.md (hunk 1 above)
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/CHANGELOG.md:6]:
```suggestion
* Added public APIs `customItemSerializer`, `getCustomItemSerializer` and `setCustomItemSerializer` to allow customers to specify custom payload transformations or serialization settings. - See [PR 38997](https://github.com/Azure/azure-sdk-for-java/pull/38997) and [PR 39933](https://github.com/Azure/azure-sdk-for-java/pull/39933) 
```

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosClientBuilder.java [hunk 3]
```diff
@@ -730,33 +731,24 @@ public CosmosClientBuilder readRequestsFallbackEnabled(boolean readRequestsFallb
...
-     * on their own.
+     * Bulk/Delete by PK/Transactional Batch/Stored Procedure execution: No automatic retries are supported.
+     * @param options the options controlling whether non-idempotent write operations should be retried and whether
+     * trackingIds can be used.
      * @return the CosmosItemRequestOptions
      */
-    public CosmosClientBuilder nonIdempotentWriteRetryPolicy(
-        boolean nonIdempotentWriteRetriesEnabled,
-        boolean useTrackingIdPropertyForCreateAndReplace) {
+    public CosmosClientBuilder nonIdempotentWriteRetryPolicy(NonIdempotentWriteRetryOptions options) {
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosClientBuilder.java:739]: Isn't this a breaking change? Am I missing something?

Comment (jeet1995) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosClientBuilder.java:739]: It was only opened up in the custom serialization PR. Not released yet.

Comment (lmolkova) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosClientBuilder.java:739]:
```suggestion
    public CosmosClientBuilder nonIdempotentWriteRetryOptions(NonIdempotentWriteRetryOptions options) {
```

for consistency

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/NonIdempotentWriteRetryOptions.java (hunk 2 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/NonIdempotentWriteRetryOptions.java:8]: Those are request options, correct? Everything used to configure ClientBuilder (SessionRetryOptions, CosmosDIagnosticsHandler, ThrottlingRetryOptions, ThroughputControl settings are all int he main package. Leaving those here.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosClientBuilder.java (hunk 3 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosClientBuilder.java:739]: Correct

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosClientBuilder.java:739]: Done

Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosClientBuilder.java]: Cool, thanks :)

**Review by xinlian12 — APPROVED** [review]
LGTM, thanks

**Pull request merged**